LNK-3702: E2E Smoke Tests

seanmcilvenna opened the pull request:
### 🛠️ Description of Changes

A smoke test embedded in the main SLN which uses Docker Compose to:
1. Loads synthetic data onto a FHIR server.
2. Loads a measure into MeasureEval and Validation service.
3. Creates/configures a tenant/facility
4. Generates an adhoc report for a pre-defined reporting period and patient
5. Waits for the report to finish generating, then downloads the report's data
6. Checks that the report was generated accurately


### 🧪 Testing Performed

The entire PR is about testing.

### 🧑‍🔬 Unit Testing
- [x] I have written or updated unit tests to cover my changes

### 📓 Documentation Updated

N/A


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Added API endpoint to download reports as ZIP archives with feature flag control.
  - Introduced Kafka message types and listeners for tracking and processing report submissions.
  - Implemented a service to generate standardized directory names for submissions and reports.
  - Added configuration improvements and updated directory structure for report storage.
  - Added reverse proxy routing for submission API with authorization.

- **Enhancements**
  - Improved error handling and logging in validation and submission consumers.
  - Enhanced Docker Compose with a FHIR server service and broader container health checks.
  - Enabled conditional endpoint access based on configuration flags.
  - Refined report generation and regeneration logic with enhanced logging.
  - Improved Kafka producer and consumer naming consistency and exception handling.

- **Testing**
  - Added comprehensive end-to-end (E2E) tests including smoke tests, data loaders, and measure loaders.
  - Added unit tests for input sanitization and supporting test projects.
  - Included embedded test data and documentation for E2E testing.

- **Chores**
  - Updated CI pipelines and solution structure to include new tests and services.
  - Added utility scripts for line ending normalization and FHIR bundle transaction conversion.

- **Bug Fixes**
  - Fixed regular expression syntax in input sanitization methods.

- **Documentation**
  - Added README files detailing E2E test project and FHIR server test data.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

seanmcilvenna committed:
Basic structure of E2E Test project

seanmcilvenna committed:
Merge branch 'dev' into tech_debt/e2e_tests

seanmcilvenna committed:
Adding unit tests to Shared project for HtmlSanitizer

seanmcilvenna committed:
Enhancements to E2E tests to create query plan and query config.

seanmcilvenna committed:
Merge branch 'dev' into tech_debt/e2e_tests

seanmcilvenna committed:
Adding CI pipeline for building and testing Shared project

seanmcilvenna committed:
Initializing validation artifacts.
Kicking off Adhoc Report request

seanmcilvenna committed:
* In submission, not getting a token for the report service when in anonymous mode.
* Updating submission configs to use ServiceRegistry instead of the old (no longer valid) submission service config's urls.
* Adding exception handling around validation kafka topic listener
* Smoke test successfully generates a report, but still need to determine how to "wait" for it to be completed and then check the results.

seanmcilvenna committed:
* Producing `ReportSubmitted` event from submission service
* Consuming `ReportSubmitted` event in report service to determine when to set `submitDate` and `submitted` properties for the scheduled report

amphillipsLGC commented on DotNet/Report/KafkaProducers/SubmitReportProducer.cs:
```diff
@@ -67,9 +67,6 @@ public async Task<bool> Produce(ReportScheduleModel schedule)
 
             _submissionReportProducer.Flush();
 
-            schedule.SubmitReportDateTime = DateTime.UtcNow;
```
This is fine, but do we want to capture a data point for when the submission was requested. Do we have a  use case for that and what was the original intent of this field.

seanmcilvenna replied:
See user story LNK-3700 which aims to improve this, later.

amphillipsLGC commented on DotNet/Report/Listeners/ReportSubmittedListener.cs:
A few things:
- It's not leveraging retry in case there is an error while updating the report scheduled entry submit date
- It's not leveraging dead letter if there is no report tracking id in the database that matches the report tracking id

seanmcilvenna replied:
I've updated the `ReportSubmittedListener` class to wrap a try/catch around the logic for handling the different error scenarios you mentioned, following similar patterns as other listeners in the report service. Please review to make sure I got it right, but it *seems* to behave as expected.
Also note, I created LNK-3701 (as low priority) to create a level of abstraction in the ReportSubmittedListener that isn't addressed by this PR.

amphillipsLGC commented on DotNet/Submission/KafkaProducers/ReportSubmittedProducer.cs:
```diff
@@ -0,0 +1,35 @@
+    public void Produce(string? correlationId, string facilityId, DateTime startDate, DateTime endDate, string reportTrackingId)
+    {
+        if (correlationId == null)
+            correlationId = Guid.NewGuid().ToString();
```
We will need to look into this if we get to continuous submission, as we would want to maintain the link between the steps. Should be fine for now though

amphillipsLGC commented on Tests/E2ETests/FhirBundleFixer.cs:
```diff
@@ -0,0 +1,75 @@
+using System.Text.Json;
+using System.Text.Json.Nodes;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
```
Can we add a comment to explain what the fixer is supposed to do.

seanmcilvenna replied:
I removed this and moved it into a Scripts/.linq file. It has an extensive comment at the top of it to explain what it does.

amphillipsLGC reviewed (approved):
Looks good!

seanmcilvenna committed:
* Altering AdhocReport endpoint to pre-determine a GUID for the report ID and return it as part of the controller endpoint so that it can be followed after it's initially generated.
* E2E unit test now able to wait for the report to be generated.

seanmcilvenna committed:
Adding rest controller to submission service to (conditionally based on config) support downloading the submission data

seanmcilvenna committed:
Updating smoke test to download the submission bundle and validate its contents

seanmcilvenna commented on Tests/E2ETests/FhirBundleFixer.cs:
```diff
@@ -0,0 +1,75 @@
+using System.Text.Json.Nodes;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+[TestClass]
+public class FhirBundleFixer
+{
```
I'm remove the FhirBundleFixer and moving it to a different utility location

seanmcilvenna commented on Tests/E2ETests/FhirBundleFixer.cs:
```diff
@@ -0,0 +1,75 @@
+[TestClass]
+public class FhirBundleFixer
+{
+    private const string BundleDirectory = @"D:\Code\link-cloud\Tests\E2ETests\data";
```
I'm remove the FhirBundleFixer and moving it to a different utility location

seanmcilvenna commented on Tests/E2ETests/FhirBundleFixer.cs:
```diff
@@ -0,0 +1,75 @@
+{
+    private const string BundleDirectory = @"D:\Code\link-cloud\Tests\E2ETests\data";
+
+    [TestMethod]
+    [Ignore]
+    public void FixFhirJsonBundlesInDirectory()
```
I'm remove the FhirBundleFixer and moving it to a different utility location

seanmcilvenna committed:
* Converting E2ETests to use XUnit.
* Moving FhirBundleFixer to a LinqPad script in the `/Scripts` folder.

seanmcilvenna committed:
Adding a README to the E2ETests project

seanmcilvenna committed:
Increasing max number of retries when waiting for the report to be submitted

seanmcilvenna committed:
Paramaterizing the LinqPad transactionize-fhir-bundle.linq script

seanmcilvenna committed:
Code quality changes

seanmcilvenna committed:
Adding LinqPad script to correct line endings in a file to be compatible with linux

seanmcilvenna committed:
Updating github action to include running e2e test

seanmcilvenna committed:
Merge branch 'dev' into tech_debt/e2e_tests

# Conflicts:
#	DotNet/Submission/Program.cs

seanmcilvenna committed:
Adding missing import/reference for Submission Program.cs

seanmcilvenna committed:
Using actions/uploa-artifact@v4 instead since v3 has been deprecated

seanmcilvenna committed:
Moving `XXXTests` projects to `Tests` solution folder

seanmcilvenna committed:
Fix for bad merge of `dev` into submission's Program.cs file.

seanmcilvenna committed:
Aligning logging messages to expected status codes

seanmcilvenna commented on Tests/E2ETests/SmokeTest.cs:
```diff
@@ -0,0 +1,371 @@
+
+        if (deleteQueryPlanResponse.StatusCode != HttpStatusCode.Accepted)
+            Console.WriteLine($"Expected HTTP 204 No Content for query plan deletion but received {deleteQueryPlanResponse.StatusCode}: {deleteQueryPlanResponse.Content}");
+    }
```
Updated the logging messages to align with the expected status codes

seanmcilvenna commented on Tests/E2ETests/SmokeTest.cs:
```diff
@@ -0,0 +1,371 @@
+
+        if (deleteQueryConfigResponse.StatusCode != HttpStatusCode.Accepted)
+            Console.WriteLine($"Expected HTTP 204 No Content for query config deletion but received {deleteQueryConfigResponse.StatusCode}: {deleteQueryConfigResponse.Content}");
+    }
```
Updated the logging messages to align with the expected status codes

seanmcilvenna commented on Tests/E2ETests/SmokeTest.cs:
```diff
@@ -0,0 +1,371 @@
+
+        if (deleteNormalizationResponse.StatusCode != HttpStatusCode.Accepted)
+            Console.WriteLine($"Expected HTTP 204 No Content for normalization deletion but received {deleteNormalizationResponse.StatusCode}: {deleteNormalizationResponse.Content}");
+    }
```
Updated the logging messages to align with the expected status codes

seanmcilvenna committed:
Handling security vulnerability in submission controller identified by CodeQL: Uncontrolled data used in path expression

seanmcilvenna committed:
Sanitizing the facility id before using

seanmcilvenna committed:
Using shared sanitize method for tenantId and reportId.
Converting sanitize methods to extension methods so they can be accessed more easily.

seanmcilvenna committed:
Implementing support of dead letter and transient exception handling in `ReportSubmittedListener`

seanmcilvenna committed:
Merge branch 'dev' into tech_debt/e2e_tests

seanmcilvenna commented on Tests/E2ETests/E2ETests.csproj:
```diff
@@ -0,0 +1,47 @@
+            <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+        </PackageReference>
+        <PackageReference Include="Hl7.Fhir.R4" Version="6.0.0-alpha1" />
+        <PackageReference Include="RestSharp" Version="112.1.1-alpha.0.4" />
```
I've changed the package versions accordingly.

seanmcilvenna commented on DotNet/Submission/KafkaProducers/ReportSubmittedProducer.cs:
```diff
@@ -0,0 +1,35 @@
+
+public class ReportSubmittedProducer(IProducer<ReportSubmittedKey, ReportSubmittedValue> producer)
+{
+    public void Produce(string? correlationId, string facilityId, DateTime startDate, DateTime endDate, string reportTrackingId)
+    {
+        if (correlationId == null)
+            correlationId = Guid.NewGuid().ToString();
+        
+        producer.Produce(nameof(KafkaTopic.ReportSubmitted), new Message<ReportSubmittedKey, ReportSubmittedValue>
+        {
+            Key = new ReportSubmittedKey()
+            {
+                FacilityId = facilityId,
+                StartDate = startDate,
+                EndDate = endDate
+            },
+            Value = new ReportSubmittedValue()
+            {
+                ReportTrackingId = reportTrackingId
+            },
+            Headers = new Headers()
+            {
+                { "X-Correlation-Id", Encoding.UTF8.GetBytes(correlationId) }
+            }
+        });
+        
+        producer.Flush();
+    }
```
Implemented as suggested

seanmcilvenna commented on DotNet/Submission/Application/Services/PathNamingService.cs:
```diff
@@ -0,0 +1,58 @@
+    public string GetSubmissionDirectoryName(string facilityId, IEnumerable<string> measures, DateTime startDate,
+        DateTime endDate, string reportId)
+    {
+        string sanitizedReportId = reportId.SanitizeAndRemove();
+        
+        //Per 2153, don't build with the trailing timestamp
+        string measureShortNames = GetMeasuresShortName(measures);
+        return $"{facilityId}-{measureShortNames}-{startDate.ToString(DirectoryDateFormat)}-{endDate.ToString(DirectoryDateFormat)}_{sanitizedReportId}";
+    }
```
I'm already sanitizing facilityId before this is called.
And the measure names are based on pre-defined service configuration.
Not going to make any changes.

seanmcilvenna commented on DotNet/Submission/Controllers/SubmissionController.cs:
```diff
@@ -0,0 +1,136 @@
+            client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", token);
+        }
+        
+        string reportUrl = $"{serviceRegistry.Value.ReportServiceApiUrl.TrimEnd('/')}/Report/summaries/{sanitizedFacilityId}?reportId={reportId}";
+        var reportResponse = await client.GetAsync(reportUrl);
+        var jsonResponse = System.Text.Json.JsonDocument.Parse(
+            await reportResponse.Content.ReadAsStringAsync());
+        
```
Implemented as suggested.

seanmcilvenna commented on DotNet/Report/Listeners/ReportSubmittedListener.cs:
```diff
@@ -0,0 +1,132 @@
+            EnableAutoCommit = false
+        };
+
+        using var consumer = kafkaConsumerFactory.CreateConsumer(config);
```
Implemented as suggested

seanmcilvenna commented on DotNet/Tenant/Controllers/FacilityController.cs:
```diff
@@ -261,7 +277,8 @@ public async Task<IActionResult> UpdateFacility(string id, FacilityConfigDto upd
         public async Task<IActionResult> DeleteFacility(string facilityId, CancellationToken cancellationToken)
         {
-            FacilityConfigModel existingFacility = _facilityConfigurationService.GetFacilityByFacilityId(facilityId, cancellationToken).Result;
+            FacilityConfigModel existingFacility = _facilityConfigurationService
+                .GetFacilityByFacilityId(facilityId, cancellationToken).Result;
```
Implemented as suggested

seanmcilvenna committed:
Implementation of CodeRabbit code-quality suggestions.

seanmcilvenna committed:
Merge remote-tracking branch 'origin/tech_debt/e2e_tests' into tech_debt/e2e_tests

seanmcilvenna commented on DotNet/Submission/Controllers/SubmissionController.cs:
```diff
@@ -0,0 +1,143 @@
+            .Select(type => type.GetString())
+            .Cast<string>().ToList();
+        
+        DateTime reportStartDate = reportStartDateElement.GetDateTime();
+        DateTime reportEndDate = reportEndDateElement.GetDateTime();
+        string reportDirectoryName =
+            pathNamingService.GetSubmissionDirectoryName(sanitizedFacilityId, reportTypes, reportStartDate, reportEndDate,
+                reportId);
```
Ignoring for this round of changes.

seanmcilvenna committed:
Sanitizing report id for URL

seanmcilvenna commented on DotNet/Submission/Controllers/SubmissionController.cs:
```diff
@@ -0,0 +1,143 @@
+        }
+
+        // TODO: Consider changing this to store the ZIP on disk, instead, and check if the ZIP already exists
+        var compressedData = this.CompressDirectory(normalizedFullPath);
```
Ignoring for this round of changes.

seanmcilvenna committed:
Merge branch 'dev' into tech_debt/e2e_tests

seanmcilvenna commented:
@coderabbit skip

seanmcilvenna commented:
@coderabbit pause

seanmcilvenna commented on DotNet/Report/Listeners/ReportSubmittedListener.cs:
```diff
@@ -0,0 +1,132 @@
+                }
+                catch (ConsumeException ex)
+                {
+                    logger.LogError(ex, "Error consuming message for topics: [{1}] at {2}",
+                        string.Join(", ", consumer.Subscription), DateTime.UtcNow);
+
+                    if (ex.Error.Code == ErrorCode.UnknownTopicOrPart)
+                    {
+                        throw new OperationCanceledException(ex.Error.Reason, ex);
+                    }
+
```
Ignoring for now.

seanmcilvenna commented on DotNet/Submission/Listeners/SubmitReportListener.cs:
```diff
@@ -172,15 +162,15 @@ await consumer.ConsumeWithInstrumentation(async (result, consumeCancellationToke
-                                string dtFormat = "yyyyMMdd";
-                                string submissionDirectory = Path.Combine(_submissionConfig.SubmissionDirectory,
-                                    $"{facilityId}-{measureShortNames}-{key.StartDate.ToString(dtFormat)}-{key.EndDate.ToString(dtFormat)}_{value.ReportTrackingId}");
+                                string submissionDirectoryName = _pathNamingService.GetSubmissionDirectoryName(
+                                    facilityId,
+                                    value.MeasureIds,
+                                    key.StartDate,
+                                    key.EndDate,
+                                    value.ReportTrackingId);
+                                string submissionDirectory = Path.Combine(
+                                    _submissionConfig.SubmissionDirectory, 
+                                    submissionDirectoryName);
 
```
Implemented as suggested

seanmcilvenna commented on DotNet/Submission/Application/Services/PathNamingService.cs:
```diff
@@ -0,0 +1,58 @@
+            .FirstOrDefault(x => x.Url == measureWithoutVersion || x.MeasureId == measureWithoutVersion)?
+            .ShortName;
+
+        if (!string.IsNullOrWhiteSpace(urlShortName))
+            return urlShortName;
+        else
+            logger.LogError("Submission service configuration does not contain a short name for measure: " + measure);
+
+        return $"{measure.GetHashCode():X}";
```
Ignoring for now.

seanmcilvenna commented on DotNet/Submission/Application/Middleware/ConditionalEndpoint.cs:
```diff
@@ -0,0 +1,46 @@
+
+                    // DownloadReport
+                    if (controllerActionDescriptor.MethodInfo.Name == "DownloadReport" && !downloadReportEnabled)
+                    {
+                        logger.LogWarning("Request to download report is rejected due to configuration.");
+                        await this.SetResponseNotEnabled(context);
+                    }
+                }
+            }
+        }
+
+        await next(context);
```
Implemented as suggested

seanmcilvenna committed:
More coderabbit fixes.

seanmcilvenna committed:
Merge branch 'dev' into tech_debt/e2e_tests

seanmcilvenna committed:
Merge remote-tracking branch 'origin/tech_debt/e2e_tests' into tech_debt/e2e_tests
